### PR TITLE
WP5 indiv process testing: code style improvements

### PR DESCRIPTION
**Pull request by soxofaan**
@m-mohr  I went through .../processing/test_example.py while playing with the individual process tests and made some minor code style tweaks along the way. 
I wanted to run them with you first before merging into main.

included tweaks:
- avoid in-place dict mutations
- add type hints
- use more keyword args for call robustness
- more clearly mark internal helpers as private
- finetune exception handling
- finetune pytest asserts (where default value rendering should be good enough)

### CODE REVIEW

File: src/openeo_test_suite/tests/processes/processing/test_example.py [hunk 1]
```diff
@@ -74,22 +89,30 @@ def test_process(
...
     # execute the process
     try:
-        result = connection.execute(id, arguments)
+        result = connection.execute(process_id, arguments)
     except Exception as e:
         result = e
 
     # check the process results / behavior
     if throws and returns:
+        # TODO what does it mean if test can both throw and return?
```
Comment (m-mohr) [inline src/openeo_test_suite/tests/processes/processing/test_example.py:115]: See process test documentation in https://github.com/Open-EO/openeo-processes/pull/468

> If both `returns` and `throws` are provided, it means that either of it should be true.
> So either the process retuens as specified or throws the specified error.

This happens e.g. for division by zero where some implementations throw an exceptions, some return ±infinity.

File: src/openeo_test_suite/tests/processes/processing/test_example.py [hunk 2]
```diff
@@ -154,110 +202,117 @@ def prepare_argument(arg, process_id, name, connection, file):
...
     elif isinstance(value, list):
         for item in value:
             check_non_json_values(item)
 
 
 def check_exception(example, result):
-    assert isinstance(result, Exception), "Excpected an exception, but got {}".format(
-        result
-    )
+    assert isinstance(result, Exception)
```
Comment (m-mohr) [inline src/openeo_test_suite/tests/processes/processing/test_example.py:290]: Why remove the message?

File: src/openeo_test_suite/tests/processes/processing/test_example.py [hunk 3]
```diff
@@ -154,110 +202,117 @@ def prepare_argument(arg, process_id, name, connection, file):
...
             )
         # todo: we should enable this end remove the two lines above, but right now tooling doesn't really implement this
         # assert result.__class__.__name__ == example["throws"]
 
 
 def check_return_value(example, result, connection, file):
-    assert not isinstance(result, Exception), "Unexpected exception: {} ".format(
-        str(result)
-    )
+    assert not isinstance(result, Exception)
```
Comment (m-mohr) [inline src/openeo_test_suite/tests/processes/processing/test_example.py:302]: Why did you remove the message?

File: src/openeo_test_suite/tests/processes/processing/test_example.py [hunk 4]
```diff
@@ -154,110 +202,117 @@ def prepare_argument(arg, process_id, name, connection, file):
...
     )
 
-    delta = example["delta"] if "delta" in example else 0.0000000001
+    delta = example.get("delta", 0.0000000001)
 
     if isinstance(example["returns"], dict):
-        assert isinstance(result, dict), "Expected a dict but got {}".format(
-            type(result)
-        )
+        assert isinstance(result, dict)
```
Comment (m-mohr) [inline src/openeo_test_suite/tests/processes/processing/test_example.py:315]: Why did you remove the message?

**Review by m-mohr — CHANGES_REQUESTED** [review]
Thanks. Some change requests, otherwise looks good in theory.
Wasn't able to run it, it looks like a recent merge corrupted the individual process tests. 
`pytest src/openeo_test_suite/tests/processes/processing/test_example.py --runner=dask` worked before, is not running any longer.

File: src/openeo_test_suite/tests/processes/processing/test_example.py (hunk 2 above)
Comment (soxofaan) [reply src/openeo_test_suite/tests/processes/processing/test_example.py:290]: because pytest by default provides a useful message already, so you don't need all this boilerplate in your tests.

e.g.

```python

def test_tmp():
    result = "lalala"
    assert isinstance(result, Exception)
```
fails with 
```
>       assert isinstance(result, Exception)
E       AssertionError: assert False
E        +  where False = isinstance('lalala', Exception)

src/openeo_test_suite/tests/test_dummy.py:22: AssertionError
```

Note how pytest mentions the value of `result` in the error report

**Comment by soxofaan** [comment]
> Wasn't able to run it, it looks like a recent merge corrupted the individual process tests.
pytest src/openeo_test_suite/tests/processes/processing/test_example.py --runner=dask worked before, is not running any longer.

what error do you get? Please create a bug ticket if that makes sense

File: src/openeo_test_suite/tests/processes/processing/test_example.py [hunk 5]
```diff
@@ -74,22 +89,30 @@ def test_process(
...
     try:
-        arguments = prepare_arguments(example["arguments"], id, connection, file)
+        arguments = _prepare_arguments(
+            arguments=example["arguments"],
+            process_id=process_id,
+            connection=connection,
+            file=file,
+        )
     except Exception as e:
-        # TODO: skipping on a generic `Exception` is very liberal and might hide real issues
```
[the remark this answers is not in the document]
Comment (m-mohr) [reply src/openeo_test_suite/tests/processes/processing/test_example.py:79]: I can only check what I meant once I can run it again ;-)

File: src/openeo_test_suite/tests/processes/processing/test_example.py (hunk 2 above)
Comment (m-mohr) [reply src/openeo_test_suite/tests/processes/processing/test_example.py:290]: I felt like the messages are easier to understand in comparison what pytest provides by default.

**Comment by soxofaan** [comment]
I restored the liberal `except: pytest.skip()` logic and moved the custom assert messages back.
Unless you have more remarks, I'd like to move this forward and merge it.
(I'd like to work on some more unification across the different WPs and that would conflict too easily with this PR I guess)

File: src/openeo_test_suite/tests/processes/processing/test_example.py [hunk 6]
```diff
@@ -154,110 +201,117 @@ def prepare_argument(arg, process_id, name, connection, file):
...
-    )
+    assert isinstance(result, Exception), f"Expected an exception, but got {result}"
     if isinstance(example["throws"], str):
         if result.__class__.__name__ != example["throws"]:
-            warnings.warn(
-                "Expected exception {} but got {}".format(
-                    example["throws"], result.__class__.__name__
-                )
+            # TODO: better way to report this warning?
+            _log.warning(
```
Comment (m-mohr) [inline src/openeo_test_suite/tests/processes/processing/test_example.py:293]: I don't get this warning anylonger in the logs, I tried before this PR and after and these ones are missing. Any clue why?

File: src/openeo_test_suite/tests/processes/processing/test_example.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (m-mohr) [reply src/openeo_test_suite/tests/processes/processing/test_example.py:79]: I'm going through it now to check what changes if we change to NotImplementedError and if the additional failures are test issues etc. Stay tuned ;-)

File: src/openeo_test_suite/tests/processes/processing/test_example.py (hunk 6 above)
Comment (soxofaan) [reply src/openeo_test_suite/tests/processes/processing/test_example.py:293]: logs of successful tests are hidden by default. 
logs of failing tests should be visible

you can disable capturing of all logs (and print statements) with options `--capture=no --log-cli-level=INFO`

Comment (m-mohr) [reply src/openeo_test_suite/tests/processes/processing/test_example.py:293]: But that shouldn't have changed. The number of successful/failing tests is the same, just the log is missing.

Comment (soxofaan) [reply src/openeo_test_suite/tests/processes/processing/test_example.py:293]: do you have an example use case (process that fails on some implementation) for this, so I can play with this too?

Comment (m-mohr) [reply src/openeo_test_suite/tests/processes/processing/test_example.py:293]: Run `pytest src/openeo_test_suite/tests/processes/processing/test_example.py --runner=dask` both on this branch and on master. 

You can then see differences:
grafik

File: src/openeo_test_suite/tests/processes/processing/test_example.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (m-mohr) [reply src/openeo_test_suite/tests/processes/processing/test_example.py:79]: I just pushed https://github.com/Open-EO/openeo-test-suite/pull/21/commits/7b28019f4ec43548ce70ed09ff71b56dfc7130d1 which switches to only NotImplementedError and fixes a couple of things in the background so that test coverage is still similar.

**Comment by soxofaan** [comment]
hmm this PR got in conflicting state with main branch.
I think I'm going to abort this, and just cherry pick some minor things

**Comment by soxofaan** [comment]
I see you merged main in this PR, but at the same time I managed to rebase PR #21 on latest main branch, and created  new PR #32, which will give  a cleaner history


I'm also considering to just cherrypick you "fix temporal issues" commit on util.py, to get that out of the way already

**Comment by m-mohr** [comment]
See #32
